### PR TITLE
Ridley Leisy - Unit 2 - Regression 1

### PR DESCRIPTION
**Pull request by RidleyLeisy**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Regression Module One - "Ridley Leisy"

# Part 1 - API Load
***
**What went well:**

- Great use of requests to obtain metrics from URL! 

```python


r = requests.get('https://wikimedia.org/api/rest_v1/metrics/pageviews/per-article/en.wikipedia/all-access/all-agents/{}/daily/{}/{}'
                 .format(wiki_page,start,end))


```

- Syntax looks great and clean:
```python

df = df[['timestamp','views']]
```

-Nice use of Regex and datetime:
```python
df['timestamp'] = pd.to_datetime(df['timestamp'],format="%Y%m%d%H")
df.head()
```
**Where you could of improved on:**

-Overall, great clean code. Try to comment more! :

```python
import numpy as np
df = df.replace('?', np.NaN)
df.isna().sum()


```


_____________________
# Quick View

***

**What went great:**

-Good job setting figure size:

```python

df.plot(figsize=(16,9))```



**-Great use of reindex :**
```python

df = df.reindex(pd.date_range(start='2015-07-01',end='2020-05-24'))

```


**Where you could of improved on:**

- Using crosstab you can achieve powerful things, combined with loops and chi-squared tests:

```python

```

# Part 3 -- Average method:

- Great job with this first assignment, and keep up the exemplary work:
```python
df['mean'] = df['views'].mean()
```

# What to work on
- [x] Practice formatting and regex
- [x] Comment more

**Comment by RidleyLeisy** [comment]
thank you!

**Comment by RidleyLeisy** [comment]
@officialbrenner today's assignment is within module 3. Along with medium post -- https://medium.com/@leisyridley/predicting-incumbent-party-vote-share-55a00bcc1b3

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# evaluating_forecasts - "Ridley Leisy"

# Part 1 - Loading Data/Analysis 
***
**What went well:**

- Good job with importing in clean, reusable code: 

```python


%matplotlib inline
import pandas as pd
import requests

project = 'en.wikipedia'
access = 'all-access'
agent = 'user'
article = 'Veganism'
granularity = 'daily'
start = '20150701'
end = '20190527'

endpoint = f'/metrics/pageviews/per-article/{project}/{access}/{agent}/{article}/{granularity}/{start}/{end}'
url = 'https://wikimedia.org/api/rest_v1' + endpoint
response = requests.get(url)
assert response.status_code == 200
df = pd.DataFrame(response.json()['items'])
df = df.rename(columns={'timestamp':'ds', 'views':'y'})
df['ds'] = pd.to_datetime(df['ds'], format='%Y%m%d%H')



```

- Nice use of datetime and I like the variable you set for the dataframe: 

```python

from fbprophet import Prophet
from fbprophet.plot import add_changepoints_to_plot

holidays = pd.DataFrame({
  'holiday': 'new_years',
  'ds': pd.to_datetime(['2015-01-01', '2016-01-01', '2017-01-01', '2018-01-01', 
                        '2019-01-01', '2020-01-01', '2021-01-01'], utc=True),
  'lower_window': -5,
  'upper_window': 5,
})

model = Prophet(daily_seasonality=False, holidays=holidays)
model.fit(df)
future = model.make_future_dataframe(periods=365*2)
forecast = model.predict(future)
fig1 = model.plot(forecast)
add_changepoints_to_plot(fig1.gca(), model, forecast)
fig2 = model.plot_components(forecast)

```

**Where you could of improved on:**

- Get into a habbit of setting a X and Y title (looks good): 


```python 

ax.set_xlabel("x label")


ax.set_ylabel("y label")

```

# Part 2 - Graphing 
***
**What went well:**

- Good job adjusting the changepoint_prior_scale parameter:

```python

import matplotlib.pyplot as plt

scales = [0.01, 0.05, 0.15, 0.50]
for scale in scales:
    model = Prophet(daily_seasonality=False, holidays=holidays, 
                    changepoint_prior_scale=scale)
    model.fit(df)
    future = model.make_future_dataframe(periods=365*2)
    forecast = model.predict(future)
    fig1 = model.plot(forecast)
    add_changepoints_to_plot(fig1.gca(), model, forecast)
    plt.title(f'changepoint_prior_scale={scale}')
    plt.show()

```

- Good showing of daily high through matplotlib(beautiful simplistic code):

```python
# TODO
ax = train.plot(x='ds',y='y',color='blue',label='train',figsize=(16,4))
test.plot(x='ds',y='y',color='red',ax=ax,label='test')
plt.title('Daily high temp in Normal, IL')


```
- Did not see any use of Prophet in the assignment so here's how you would plot the mean absolute error .:

```python


from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mae')




```

# What to work on
- [x] Comment more in code 
- [x] Add X and Y labels 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Ridley Leisy- "Regression 1 Sprint Challenge"

# Part 1. Predict avocado sales
***
**What went well:**

-Good job imporing data and dataset and checking out shape:

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)

```

## 1.2. Plot the forecast

- Took it a step further with Prophet so I like this code (added simpler line for future case below):
```python


avocado.columns = ['ds','y']



model.plot(preds)




```

- Using Prophet: 

```python


from fbprophet import Prophet
avocado.columns = ['ds', 'y']
m = Prophet()
m.fit(avocado)
future = m.make_future_dataframe(periods=365*2)
forecast = m.predict(future)

#one liner
fig1 = m.plot(forecast)


```




```

##1.5. Plot the cross-validation Mean Absolute Percentage Error: metric='mape' 
***
**What went well:**

- Clean and elegant :

```python


fig1 = m.plot(forecast)

```


** What could be improved: 

```python 
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mape')

#This is what we use 
```

#Stretch Goal (PARTIAL)

# Part 2 

##2.1. Begin with baselines for regression. Calculate the mean absolute error and 𝑅2 score for a mean baseline.

- Went went the base .head: 


``` python


rent.head()





```

**What could be better

-use this syntax for better insight:
```python

rent.describe()

```


## 2.2. Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
-Good job choosing features and getting to it (I added another way as well)):
```python

#Your Code
X = rent[['bedrooms','bathrooms']]
y = rent['price']


features = ['bedrooms','bathrooms']

target = 'price'

X = rent[features]
y = rent[target]

model.fit(X, y)

y_pred = model.predict(X)



train = rent.query('month <= 5')
test  = rent.query('month == 6')
X_train = train[features] 
y_train = train[target]
X_test  = test[features]
y_test  = test[target]


model.fit(X_train, y_train)


y_pred = model.predict(X_test)



#My Syntax

features = ['bedrooms']
target = 'price'
x = rent[features]
y = rent[target]


## Other option

train = rent.query('month < 6')
test = rent.query('month == 6')
train.shape, test.shape

features = ['bedrooms', 'bathrooms']
target = 'price'

```

##2.3. Choose any two features.

-use this syntax for better insight:
```python

```

2.4. Use scikit-learn to fit a Linear Regression model on the train data.

-Not bad for just starting out to use to scikit. Syntax is not bad, but features could be better ( see below ):
```python

from sklearn.metrics import mean_absolute_error, r2_score
import matplotlib.pyplot as plt

def plot_preds(x, y, y_pred):
    plt.scatter(x, y, label='y_true')
    plt.plot(x, y_pred, label='y_pred')
    plt.legend()
    
    
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    print('Mean Absolute Error:', mae)
    print('R^2:', r2)
    
    
    
plot_preds(x, y, y_pred)
plt.title('Mean Baseline');

```

**What could be better

-better syntax:


```python

# Choose a class of model by importing the appropriate estimator class from Scikit-Learn.
from sklearn.linear_model import LinearRegression

# Choose model hyperparameters by instantiating this class with desired values.
model = LinearRegression()

# Arrange data into a features matrix and target vector.
X_train = train[features]
y_train = train[target]
X_test = test[features]
y_test = test[target]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

# Fit the model to your data by calling the fit() method of the model instance.
model.fit(X_train, y_train)

# Apply the Model to new data: For supervised learning, often we predict labels for unknown data using the predict() method.
y_pred = model.predict(X_test)

```

2.6. Get regression metrics MAE and 𝑅2 for the test data.

-Great job:
```python


mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('Test Mean Absolute Error:', mae)
print('Test R^2 Score:', r2)

```

2.7. Get the model's coefficients and intercept.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

-use this syntax for better insight:
```python
model.intercept_, model.coef_
```

#Stretch Goals_2

```python

import seaborn as sns

target = 'price'
features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
for feature in features:
    sns.lmplot(x=feature, y=target, data=train, scatter_kws={'alpha':0.01})
    
    ```
    
    
    
    ```python
    
    
    def fit(train, test, features):
    X_train = train[features]
    y_train = train[target]
    X_test = test[features]
    y_test = test[target]

    model = LinearRegression()
    model.fit(X_train, y_train)

    print(f'Linear Regression with {len(features)} features: {features}')
    print('Train Mean Absolute Error:', 
          mean_absolute_error(y_train, model.predict(X_train)))
    print('Train R^2 Score:', 
          r2_score(y_train, model.predict(X_train)))
    print('Test Mean Absolute Error:', 
          mean_absolute_error(y_test, model.predict(X_test)))
    print('Test R^2 Score:', 
          r2_score(y_test, model.predict(X_test)))

features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
fit(train, test, features)
```

-with 3 features:

```python
features = ['bedrooms', 'bathrooms', 'longitude']
fit(train, test, features)
```

# What to work on
- [x] Comment more in code 
- [x] Less lines and more organized notebooks 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Ridley Leisy
## Assignment: Doing Linear Regression[election_regression]


# Part 1. Imports

***
**What went well:**

-Good collective use of imports. Nice and clean code for presidents:

```python
import pandas as pd
import numpy as np

columns = ['Year','Incumbent Party Candidate','Other Candidate','Incumbent Party Vote Share']

data = [[1952,"Stevenson","Eisenhower",44.6],
        [1956,"Eisenhower","Stevenson",57.76],
        [1960,"Nixon","Kennedy",49.91],
        [1964,"Johnson","Goldwater",61.34],
        [1968,"Humphrey","Nixon",49.60],
        [1972,"Nixon","McGovern",61.79],
        [1976,"Ford","Carter",48.95],
        [1980,"Carter","Reagan",44.70],
        [1984,"Reagan","Mondale",59.17],
        [1988,"Bush, Sr.","Dukakis",53.94],
        [1992,"Bush, Sr.","Clinton",46.55],
        [1996,"Clinton","Dole",54.74],
        [2000,"Gore","Bush, Jr.",50.27],
        [2004,"Bush, Jr.","Kerry",51.24],
        [2008,"McCain","Obama",46.32],
        [2012,"Obama","Romney",52.00], 
        [2016,"Clinton","Trump",48.2]]
        
votes = pd.DataFrame(data=data, columns=columns)

```
# Wrangling
-Sexy code! (Comment your notebook)   : 

```python
grouped = ted.groupby(by=[ted['DATE'].dt.year,ted['DATE'].dt.month]).mean()
g = pd.merge(grouped.xs(8,level=1),grouped.xs(9,level=1),left_index=True,right_index=True)
g = pd.merge(g,grouped.xs(10,level=1),left_index=True,right_index=True)      

```

- Good use of [.OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html), makes things easier to work with! (Also nice addition of remembering the setting of the axis:

  
```python
grouped = ted.groupby(by=[ted['DATE'].dt.year,ted['DATE'].dt.month]).mean()
g = pd.merge(grouped.xs(8,level=1),grouped.xs(9,level=1),left_index=True,right_index=True)
g = pd.merge(g,grouped.xs(10,level=1),left_index=True,right_index=True)

g = g.mean(axis=1)
```

# Predicting
## Imports

- Nice import of sklearn a very import library you will find in your career (before it's outdated): 

```python


from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

```

# What to work on
- [x] Comment more in code
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Ridley Leisy- Regression 1 Sprint Challenge

# Part 1. Predict avocado sales

***
**What went well:**

-Good collective use of imports and reading in data of the :

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)

```
# Part 2. Predict NYC apartment rent

-Good importation of data :

```python
import pandas as pd
rent_url = 'https://drive.google.com/uc?export=download&id=1lVVPmh-WYGb_0Gs_lOv22i02-150qDdZ'
rent = pd.read_csv(rent_url)
assert rent.shape == (48295, 7)

```

- importing stretch +1 : 

```python
from sklearn.metrics import mean_absolute_error, r2_score, mean_squared_error
```

## Importing fbprophet

- Great job bringing in Facebook's Prophet API and using .fit on the avocado_cv dataset: 

```python
model = Prophet()
model.fit(avocado)
```

- Good predictions:

```python
model.plot(preds)
```



## Prophet Plotting

- Very simple but powerful code! Nice use of comments too :

```python


fig1 = m.plot(forecast)




# Plot components
fig2 = m.plot_components(forecast)



```

## Calculating MAPE(Mean Absolute Percentage Error)

- Impressive use of finding the MAPE!:

```python
# mean baseline error
mean_absolute_error(rent['price'],rent['mean_baseline'])

mean_absolute_error(rent['price'],rent['naive_baseline'])

r2_score(rent['price'],rent['mean_baseline'])

r2_score(rent['price'],rent['naive_baseline'])

from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(avocado_cv, metric='mae')


```

**What could be better**

## R^2 score

- Code is not invalid but could be cleaner like so:
```python
print('The MAE for test is:', rent_test_mae, '\n')
print('The R^2 for test is:', rent_test_r2)
```
# Part 2. Predict NYC apartment rent

## Setting Target

- Nice job setting feature and target of rent:  

```python
features = ['bedrooms','bathrooms']
target = 'price'

```

## MAE

- Excellent work. I mean this because this is important when looking at features:

```python
mean_absolute_error(test['price'],preds)
```
- Nice job getting r^2 and y hat out there to work:

```python

r2_score(test['price'],preds)
```
#Stretch Goal

- DM me 

```python
:) 

```


#What to work on
- [x] Comment more in code
- [x] Come to PM for questions or help
